pkg/cmd/roachtest/registry/operation_spec.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package registry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/operation"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+)
+
+type OperationDependency int
+
+const (
+	OperationRequiresNodes OperationDependency = iota
+	OperationRequiresSqlConnection
+	OperationRequiresDatabaseSchema
+	OperationRequiresPopulatedDatabase
+)
+
+// OperationSpec is a spec for a roachtest operation.
+type OperationSpec struct {
+	Skip string // if non-empty, operation will be skipped
+	// When Skip is set, this can contain more text to be printed in the logs
+	// after the "--- SKIP" line.
+	SkipDetails string
+
+	Name string
+	// Owner is the name of the team responsible for signing off on failures of
+	// this test that happen in the release process. This must be one of a limited
+	// set of values (the keys in the roachtestTeams map).
+	Owner Owner
+	// The maximum duration the test is allowed to run before it is considered
+	// failed. If not specified, the default timeout is 10m before the test's
+	// associated cluster expires. The timeout is always truncated to 10m before
+	// the test's cluster expires.
+	Timeout time.Duration
+
+	// CompatibleClouds is the set of clouds this test can run on (e.g. AllClouds,
+	// OnlyGCE, etc). Must be set.
+	CompatibleClouds CloudSet
+
+	// Suites is the set of suites this test is part of (e.g. Nightly, Weekly,
+	// etc). Must be set, even if empty (see ManualOnly).
+	Suites SuiteSet
+
+	// Dependency specify the types of resources required for this roachtest
+	// operation to work. This will be used to match this operation up with
+	// eligible clusters to run.
+	//
+	// TODO(bilal): Unused.
+	Dependency OperationDependency
```

Comment:
Should we introduce these when they have a meaning? It's confusing having to set this given that the semantics are undefined (same for other unused fields in this spec).

Reply:
The key purpose of this PR is to introduce an API and allow other teams to be able to write operations ASAP. It's the only way to improve DRT's effectiveness as we're a very small team that can't write operations for every single team. It's why I want to get some way of defining dependencies of an operation out, even if we end up stripping it out later.

Reply:
I want to get some way of defining dependencies of an operation out, so operation authors have a complete-enough API. Stripping it out later (if we choose to do that) is easier than going back and adding a new parameter. But I can remove it if it's particularly in the way.

Reply:
The idea is that these are hierarchical (I think I replied as such in the reviewable thread but it got lost here). Populated DB implies schema, implies SQL connection, implies existent nodes (which is the 0 value). Also added that as a comment on the type above. See `TestFilter` in #119796 for how this could be used. Is that sufficient? I don't mind removing it (drt-run isn't using it). But that was the original idea.